File: CoreLib/System/Threading/WaitHandle.cs
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Apache License 2.0 (Apache)
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
namespace System.Threading
{
    using System.Threading;
    using System.Runtime.CompilerServices;
    using System;
    public abstract class WaitHandle : MarshalByRefObject
    {
        public const int WaitTimeout = 0x102;
        
        public virtual bool WaitOne(int millisecondsTimeout, bool exitContext)
        {
            throw new NotImplementedException();
        }

        public virtual bool WaitOne()
        {
            return WaitOne(Timeout.Infinite, false);
        }

        
        private static int WaitMultiple(WaitHandle[] waitHandles, int millisecondsTimeout, bool exitContext, bool WaitAll)
        {
            throw new NotImplementedException();
        }

        public static bool WaitAll(WaitHandle[] waitHandles, int millisecondsTimeout, bool exitContext)
        {
            return (WaitMultiple(waitHandles, millisecondsTimeout, exitContext, true /* waitall*/ ) != WaitTimeout);
        }

        public static bool WaitAll(WaitHandle[] waitHandles)
        {
            return WaitAll(waitHandles, Timeout.Infinite, true);
        }

        public static int WaitAny(WaitHandle[] waitHandles, int millisecondsTimeout, bool exitContext)
        {
            return WaitMultiple(waitHandles, millisecondsTimeout, exitContext, false /* waitany*/ );
        }

        public static int WaitAny(WaitHandle[] waitHandles)
        {
            return WaitAny(waitHandles, Timeout.Infinite, true);
        }
    }
}

File: Il2Native.Logic/NativeImplementations/CoreLib/System/Threading/WaitHandle.cpp
#include "CoreLib.h"

namespace CoreLib {
	namespace System {
		namespace Threading {

			// Method : System.Threading.WaitHandle.WaitOneNative(System.Runtime.InteropServices.SafeHandle, uint, bool, bool)
			int32_t WaitHandle::WaitOneNative(::CoreLib::System::Runtime::InteropServices::SafeHandle* waitableSafeHandle, uint32_t millisecondsTimeout, bool hasThreadAffinity, bool exitContext)
			{
#ifndef GC_PTHREADS
				return ::WaitForSingleObject((HANDLE)waitableSafeHandle->handle.ToInt32(), millisecondsTimeout);
#else
#error NOT IMPLEMENTED YET
#endif
			}

			// Method : System.Threading.WaitHandle.WaitMultiple(System.Threading.WaitHandle[], int, bool, bool)
			int32_t WaitHandle::WaitMultiple(__array<WaitHandle*>* waitHandles, int32_t millisecondsTimeout, bool exitContext, bool WaitAll)
			{
#ifndef GC_PTHREADS
				return ::WaitForMultipleObjects(waitHandles->_length, (HANDLE*)&waitHandles->_data[0], WaitAll ? 1 : 0, millisecondsTimeout);
#else
#error NOT IMPLEMENTED YET
#endif
			}

			// Method : System.Threading.WaitHandle.SignalAndWaitOne(Microsoft.Win32.SafeHandles.SafeWaitHandle, Microsoft.Win32.SafeHandles.SafeWaitHandle, int, bool, bool)
			int32_t WaitHandle::SignalAndWaitOne(::CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle* waitHandleToSignal, ::CoreLib::Microsoft::Win32::SafeHandles::SafeWaitHandle* waitHandleToWaitOn, int32_t millisecondsTimeout, bool hasThreadAffinity, bool exitContext)
			{
				throw 0xC000C000;
			}
		}
	}
}